cachito/workers/pkg_managers/pip.py

```diff
@@ -1575,3 +1578,104 @@ def get_hosted_repositories_username(request_id):
+        for the request
+    :param str username: the username of the unprivileged user that has read access
+        to the main Cachito repositories
+    :param str password: the password of the unprivileged user that has read access
+        to the main Cachito repositories
+    :param str custom_ca_path: the path to set ``cert`` to in the pip.conf file; if not provided,
+        this option will be omitted
+    :return: the contents of the pip.conf file
+    :rtype: str
+    """
+    # Instead of getting the token from Nexus, use basic authentication as supported by Pip:
+    # https://pip.pypa.io/en/stable/user_guide/#basic-authentication-credentials
+    # Create the request specific nexus repo url to be passed as `index-url` within pip.conf
+    if "://" in nexus_pypi_hosted_repo_url:
+        nexus_pypi_hosted_repo_url = nexus_pypi_hosted_repo_url.replace(
+            "://", f"://{username}:{password}@", 1
+        )
+    else:
+        raise CachitoError(
+            f"Nexus PyPI Hosted Repo URL: {nexus_pypi_hosted_repo_url} is not a valid URL"
+        )
+
+    pip_conf_content = textwrap.dedent(
+        f"""\
+        [global]
+        index-url = {nexus_pypi_hosted_repo_url}
+        """
+    )
+    if custom_ca_path:
+        # The CA could be embedded in the actual pip.conf file with the `cert` option, but this
+        # would mean that the CA file contents would be duplicated in the Cachito database for
+        # every request that uses a pip conf file
+        pip_conf_content += f'cert = "{custom_ca_path}"\n'
+
+    return pip_conf_content
+
+
+def create_custom_pip_conf_file(request_id, package_configs=None):
+    """
+    Create a custom pip.conf file.
+
+    It should configure pip to point to the request specific repository in Nexus.
+
+    :param int request_id: the ID of the Cachito request
+    :param list package_configs: the list of optional package configurations submitted by
+    the user
+    :rtype: list<dict>
+    :return: The list of files for the pip configuration files that point to the request
+    specific repository in Nexus
+    """
+    if package_configs is None:
+        package_configs = []
+
+    config = get_worker_config()
+    nexus_pypi_hosted_repo_url = get_pypi_hosted_repo_url(request_id)
+    ca_cert = nexus.get_ca_cert()
+    subpaths = [os.path.normpath(c["path"]) for c in package_configs if c.get("path")]
+    if not subpaths:
+        # Default to the root of the application source
+        subpaths = [os.curdir]
+
+    log.info("Generating the pip conf files")
+    pip_config_files = []
+    if ca_cert:
+        # The custom CA will be called ca.cert in the "app" directory
+        pip_config_files.append(
+            {
+                "content": base64.b64encode(ca_cert.encode("utf-8")).decode("utf-8"),
+                "path": os.path.join("app", "ca.cert"),
+                "type": "base64",
+            }
+        )
+
+    for subpath in subpaths:
```

Comment:
Do we need to generate a config file for each of the subpaths? Unlike `.npmrc`, the location of `pip.conf` should not matter as long as we set the `PIP_CONFIG_FILE` environment variable. I believe we need to set the variable anyway, unless we can somehow put the file in one of the default locations, see https://pip.pypa.io/en/stable/user_guide/#config-file

I think what we need is the cert, a global (for the request) `pip.conf` file and a `PIP_CONFIG_FILE` env var

Reply:
I'm not sure how will we be configuring the `PIP_CONFIG_FILE` env var here to have it all accessible but according to the Cachito Python Support doc: "multiple paths should be supported like the “npm” package manager." and hence, we are generating config file for each of the sub paths.

Reply:
There is a difference between `npm` and `pip` when it comes to handling config files. When executing `npm`, it will look for `<current_directory>/.npmrc`. Pip, however, will *not* look for `<current_directory>/pip.conf`, so even if we generate one such file for each subpath, they will all be ignored. 

If I'm not mistaken, the only difference between the N generated pip.conf files for each of the subpaths would be the `cert` value, which is a relative path from the subdirectory to the CA cert. If that's the case, then we only need one pip.conf with the correct cert path, and we need to correctly set the `PIP_CONFIG_FILE` environment variable.

Reply:
Continued in https://github.com/release-engineering/cachito/pull/248#issuecomment-684902738